dss/operations/secrets.py

```diff
@@ -0,0 +1,280 @@
+"""
+Get/set secret variable values from the AWS Secrets Manager
+"""
+import os
+import sys
+import select
+import typing
+import argparse
+import json
+import logging
+
+from dss.operations import dispatch
+from dss.util.aws.clients import secretsmanager  # type: ignore
+
+logger = logging.getLogger(__name__)
+
+
+def get_secretsmanager_prefix(args):
+    """
+    Use information from the environment to assemble
+    the necessary prefix for accessing variables in
+    the SecretsManager.
+    """
+    store_name = os.environ["DSS_SECRETS_STORE"]
+    if args.stage is None:
+        stage_name = os.environ["DSS_DEPLOYMENT_STAGE"]
+
+    store_prefix = f"{store_name}/{stage_name}/"
+    return store_prefix
+
+
+events = dispatch.target("secrets", arguments={}, help=__doc__)
+
+
+@events.action(
+    "list",
+    arguments={
+        "--stage": dict(
+            required=False, help="the stage for which secrets should be listed"
+        ),
+        "--json": dict(
+            default=False,
+            action="store_true",
+            help="format the output as JSON if this flag is present",
+        ),
+    },
+)
+def list_secrets(argv: typing.List[str], args: argparse.Namespace):
+    """
+    Print a list of names of every secret variable in the secrets manager
+    for the given store and stage.
+    """
+    store_prefix = get_secretsmanager_prefix(args)
+
+    paginator = secretsmanager.get_paginator("list_secrets")
+
+    secret_names = []
+    for response in paginator.paginate():
+        for secret in response["SecretList"]:
+
+            # Get resource IDs
+            secret_name = secret["Name"]
+
+            # Only save secrets for this store and stage
+            if secret_name.startswith(store_prefix):
+                secret_names.append(secret_name)
+
+    secret_names.sort()
+
+    if args.json:
+        print(json.dumps(secret_names))
+    else:
+        for secret_name in secret_names:
+            print(secret_name)
+
+
+@events.action(
+    "get",
+    arguments={
+        "--stage": dict(
+            required=False, help="the stage for which secrets should be listed"
+        ),
+        "--secret-name": dict(
+            required=True,
+            nargs="*",
+            help="name of secret or secrets to retrieve (list values can be separated by a space)",
+        ),
+        "--json": dict(
+            default=False,
+            action="store_true",
+            help="format the output as JSON if this flag is present",
+        ),
+    },
+)
+def get_secret(argv: typing.List[str], args: argparse.Namespace):
```

Comment:
Maybe I'm being overly paranoid, but we should never return the encrypted secret with operation tasks. Especially if these operation will be running mostly our local machines. on At most we should be able to list and set. If we need to view key which should be rare, it should be done through the AWS console.

Reply:
This is intended to replace `scripts/fetch-secret.py`, which we use programmatically in our CI/CD scripting.

Reply:
@Bento007 I agree we need to think hard about whether the get functionality is necessary, and that the AWS console is the optimal place to be viewing secret values. It is a bit too much power to be able to `print all the secrets`. 

@xbrianh this does beg the question - where is the need for this functionality to get the value of a secret in the CI/CD scripting coming from, and can we address that by removing the need somehow? In a best case scenario, it would be as simple as replacing shell scripts with Python scripts to eliminate the need to pass secrets around and use the secrets manager directly.

Reply:
I don't think there is an easy way to remove the need to pull secrets for our deployments. This prevents secrets from being stored on an operators machine, which would be a bigger threat target..

Reply:
Operators can already dump secrets to STDOUT with `aws secretsmanager get-secret-value`. The current (and future) scripting facilitates the identification and usage of DSS secrets.

The CI/CD jobs require secrets for authorization to access infrastructure during testing, and to install authorization keys into deployed DSS services.

Reply:
This can be a future feature, but we should fetch the secret in the lambda rather than baking in during deployment. This will allow us to rotate the secret with out requiring a redeployment.

In general this is an improvement to operations and security compared to what we currently have.

Reply:
@Bento007 There are tradeoffs to calling secretsmanager from deployed Lambdas. This is something we should consider, but each network call reduces the reliability and performance of our Lambdas in ways that are not under our control.

Reply:
The secret can be retrieve once per lambda when the lambda is spun up. This would only hurt the reliability of starting a new lambda which only happens when we scale horizontally. I think it's worth exploring in the future.